contrib/platform/src/com/sun/jna/platform/win32/OaIdl.java

```diff
@@ -1687,18 +1685,34 @@ public HREFTYPE getHreftype() {
         public VARTYPE vt;
 
         public TYPEDESC() {
-            this.read();
         }
 
         public TYPEDESC(Pointer pointer) {
             super(pointer);
-            this.read();
         }
 
         public TYPEDESC(_TYPEDESC _typedesc, VARTYPE vt) {
             this._typedesc = _typedesc;
             this.vt = vt;
         }
+
+        @Override
+        public void read() {
+            readField("vt");
+            switch (vt.intValue()) {
+                case Variant.VT_PTR :
+                case Variant.VT_SAFEARRAY :
+                    _typedesc.setType("lptdesc");
+                    break;
+                case Variant.VT_CARRAY :
+                    _typedesc.setType("lpadesc");
+                    break;
+                case Variant.VT_USERDEFINED :
```

Comment:
Should this also be the default? the `hreftype` is safe to read in any situation as it is just a plain `DWORD` and thus won't cause further reads. For the `VARDESC` this is implicitly done and I think it would be right here too.

Reply:
It may be safe to read, but it sets the union's type to a value that doesn't apply. I consider this to be incorrect. But maybe my idea of unions is wrong.

Reply:
The problem I see is, that if you don't set the type, it might be already be set to one of the other types, which would result in an invalid read. It is the responsibility of the calling code to only read the value of the union if it makes sense, but we need to protect from invalid reads.